File: f77/build08/kmtc.f
C
        Subroutine Kinematics_Data
C
C       This subroutine sets the Default data for the kinematics routine.
C
C       Inputs:
C           None.
C
C       Outputs:
C           Gravity                 - Acceleration due to gravity [m/sec**2]
C           Theta_b_IC_dg   - Initial attitude of missile [deg]
C           Q_b_IC_dg               - Initial attitude rate of missile [deg/sec]
C           X_bi_i_IC               - Initial X position of BCS w.r.t. ICS, expressed in the
C                                             ICS [m]
C           Xd_bi_i_IC              - Initial X velocity of BCS w.r.t. ICS, expressed
C                                             ICS [m/sec]
C           Z_bi_i_IC               - Initial Z position of BCS w.r.t. ICS, expressed
C                                             ICS [m]
C           Zd_bi_i_IC              - Initial Z velocity of BCS w.r.t. ICS, expressed
C                                             ICS [m/sec]
C
C       Internal variables and constants:
C           None.
C
C       Declare global common and assign variable locations
C
      Include '../driver/global.inc'
      Include '../driver/sysvars.inc'
      Equivalence (Real_Array(509), X_bi_i_IC)
      Equivalence (Real_Array(510), Xd_bi_i_IC)
      Equivalence (Real_Array(511), Z_bi_i_IC)
      Equivalence (Real_Array(512), Zd_bi_i_IC)
      Equivalence (Real_Array(513), Theta_b_IC_dg)
      Equivalence (Real_Array(514), Q_b_IC_dg)
      Equivalence (Real_Array(515), Gravity)
C
C       Begin default data definitions:
C
      Theta_b_IC_dg =    0.0e0
      Q_b_IC_dg     =    0.0e0
      X_bi_i_IC     = -500.0e0
      Z_bi_i_IC     = -100.0e0
      Xd_bi_i_IC    =  100.0e0
      Zd_bi_i_IC    =    0.0e0
      Gravity       =    9.88e0
C
      Return
      End

      Subroutine Kinematics_Init
C
C       This subroutine initializes the variables for the kinematics routine.
C
C       Inputs:
C           RDTODG        - Radians to degrees conversion factor [deg/rad]
C           Theta_b_IC_dg - Initial attitude of missile [deg]
C           Q_b_IC_dg     - Initial attitude rate of missile [deg/sec]
C           X_bi_i_IC     - Initial X position of BCS w.r.t. ICS, expressed in the
C                           ICS [m]
C           Xd_bi_i_IC    - Initial X velocity of BCS w.r.t. ICS, expressed
C                           ICS [m/sec]
C           Z_bi_i_IC     - Initial Z position of BCS w.r.t. ICS, expressed
C                           ICS [m]
C           Zd_bi_i_IC    - Initial Z velocity of BCS w.r.t. ICS, expressed
C                           ICS [m/sec]
C
C       Outputs:
C           Theta_b       - Attitude of missile [rad]
C           Q_b           - Attitude rate of missile [rad/sec]
C           X_bi_i        - X position of BCS w.r.t. ICS, expressed in the ICS [m]
C           Xd_bi_i       - X velocity of BCS w.r.t. ICS, expressed ICS [m/sec]
C           Z_bi_i        - Z position of BCS w.r.t. ICS, expressed ICS [m]
C           Zd_bi_i       - Z velocity of BCS w.r.t. ICS, expressed ICS [m/sec]
C
C       Internal variables and constants:
C           None.
C
C       Declare global common and assign variable locations
C
      Include '../driver/global.inc'
      Include '../driver/sysvars.inc'
      Equivalence (Real_Array( 50), RDTODG)
      Equivalence (Real_Array(500), X_bi_i)
      Equivalence (Real_Array(501), Xd_bi_i)
      Equivalence (Real_Array(503), Z_bi_i)
      Equivalence (Real_Array(504), Zd_bi_i)
      Equivalence (Real_Array(506), Theta_b)
      Equivalence (Real_Array(507), Q_b)
      Equivalence (Real_Array(509), X_bi_i_IC)
      Equivalence (Real_Array(510), Xd_bi_i_IC)
      Equivalence (Real_Array(511), Z_bi_i_IC)
      Equivalence (Real_Array(512), Zd_bi_i_IC)
      Equivalence (Real_Array(513), Theta_b_IC_dg)
      Equivalence (Real_Array(514), Q_b_IC_dg)

C       Begin math model initialization
C
C       Define the states
C
      Call Define_Real_State(500, 501)
      Call Define_Real_State(501, 502)
      Call Define_Real_State(503, 504)
      Call Define_Real_State(504, 505)
      Call Define_Real_State(506, 507)
      Call Define_Real_State(507, 508)
C
C       Set the initial conditions
C
      X_bi_i          = X_bi_i_IC
      Z_bi_i          = Z_bi_i_IC
      Xd_bi_i         = Xd_bi_i_IC
      Zd_bi_i         = Zd_bi_i_IC
      Theta_b         = Theta_b_IC_dg/RDTODG
      Q_b             = Q_b_IC_dg/RDTODG
C
      Return
      End

      Subroutine Kinematics
C
C       This subroutine determined the kinematic state derivatives.
C
C       Inputs:
C           AaeroX_bi_b - Accelerations due to aero in BCS [mXsec**2)
C           AaeroZ_hi_b
C           Gravity - Acceleration due to gravity [m/sec**2]
C           Theta_b - Attitude of missile [rad]
C
C       Outputs:
C           Xdd_bi_i - X acceleration of BCS w.r.t. ICS, expressed in the ICS [m/sec]
C           Zdd_bi_i - Z acceleration of BCS w.r.t. ICS expressed in the ICS [m/sec]
C
C       Internal variables and constants:
C           None.
C
C       Declare global common and assign variable locations
C
      Include '../driver/global.inc'
      Include '../driver/sysvars.inc'
      Equivalence (Real_Array(100), AaeroX_bi_b)
      Equivalence (Real_Array(101), AaeroZ_bi_b)
      Equivalence (Real_Array(502), Xdd_bi_i)
      Equivalence (Real_Array(505), Zdd_bi_i)
      Equivalence (Real_Array(506), Theta_b)
      Equivalence (Real_Array(515), Gravity)
C
C       Begin math model:
C
C       Calculate accelerations (attitude rate derivative comes from aero)
C
      Xdd_bi_i =  AaeroX_bi_b*Cos(Theta_b) + AaeroZ_bi_b*Sin(Theta_b)
      Zdd_bi_i = -AaeroX_bi_b*Sin(Theta_b) + AaeroZ_bi_b*Cos(Theta_b)+
     &               Gravity
C
      Return
      End

File: compcert/build08/kmtc.c
#include "kmtc.h"
#include "../driver/global.h"
#include "../driver/sysvars.h"
#include "../driver/deriv.h"
#include "../driver/defstate.h"
#include <math.h>

#define aaerox_bi_b   real_array[100].r
#define aaeroz_bi_b   real_array[101].r
#define xdd_bi_i      real_array[502].r
#define zdd_bi_i      real_array[505].r
#define theta_b       real_array[506].r
#define gravity       real_array[515].r
#define rdtodg        real_array[50].r
#define x_bi_i        real_array[500].r
#define xd_bi_i       real_array[501].r
#define z_bi_i        real_array[503].r
#define zd_bi_i       real_array[504].r
#define q_b           real_array[507].r
#define x_bi_i_ic     real_array[509].r
#define xd_bi_i_ic    real_array[510].r
#define z_bi_i_ic     real_array[511].r
#define zd_bi_i_ic    real_array[512].r
#define theta_b_ic_dg real_array[513].r
#define q_b_ic_dg     real_array[514].r

void kinematics_data() {
  // This subroutine sets the Default data for the kinematics routine.
  //
  // Inputs:
  //     None.
  //
  // Outputs:
  //     Gravity       - Acceleration due to gravity [m/sec**2]
  //     Theta_b_IC_dg - Initial attitude of missile [deg]
  //     Q_b_IC_dg     - Initial attitude rate of missile [deg/sec]
  //     X_bi_i_IC     - Initial X position of BCS w.r.t. ICS, expressed in the ICS [m]
  //     Xd_bi_i_IC    - Initial X velocity of BCS w.r.t. ICS, expressed ICS [m/sec]
  //     Z_bi_i_IC     - Initial Z position of BCS w.r.t. ICS, expressed ICS [m]
  //     Zd_bi_i_IC    - Initial Z velocity of BCS w.r.t. ICS, expressed ICS [m/sec]
  // Begin default data definitions:
  theta_b_ic_dg = 0.0;
  q_b_ic_dg = 0.0;
  x_bi_i_ic = -500.0;
  z_bi_i_ic = -100.0;
  xd_bi_i_ic = 100.0;
  zd_bi_i_ic = 0.0;
  gravity = 9.88;
}

void kinematics_init() {
  // Inputs:
  //     RDTODG        - Radians to degrees conversion factor [deg/rad]
  //     Theta_b_IC_dg - Initial attitude of missile [deg]
  //     Q_b_IC_dg     - Initial attitude rate of missile [deg/sec]
  //     X_bi_i_IC     - Initial X position of BCS w.r.t. ICS, expressed in the
  //                     ICS [m]
  //     Xd_bi_i_IC    - Initial X velocity of BCS w.r.t. ICS, expressed
  //                     ICS [m/sec]
  //     Z_bi_i_IC     - Initial Z position of BCS w.r.t. ICS, expressed
  //                     ICS [m]
  //     Zd_bi_i_IC    - Initial Z velocity of BCS w.r.t. ICS, expressed
  //                     ICS [m/sec]
  //
  // Outputs:
  //     Theta_b       - Attitude of missile [rad]
  //     Q_b           - Attitude rate of missile [rad/sec]
  //     X_bi_i        - X position of BCS w.r.t. ICS, expressed in the ICS [m]
  //     Xd_bi_i       - X velocity of BCS w.r.t. ICS, expressed ICS [m/sec]
  //     Z_bi_i        - Z position of BCS w.r.t. ICS, expressed ICS [m]
  //     Zd_bi_i       - Z velocity of BCS w.r.t. ICS, expressed ICS [m/sec]
  // Begin math model initialization:
  // Define the states
  define_real_state(500,501);
  define_real_state(501,502);
  define_real_state(503,504);
  define_real_state(504,505);
  define_real_state(506,507);
  define_real_state(507,508);
  // Set the initial conditions
  x_bi_i  = x_bi_i_ic;
  z_bi_i  = z_bi_i_ic;
  xd_bi_i = xd_bi_i_ic;
  zd_bi_i = zd_bi_i_ic;
  theta_b = theta_b_ic_dg/rdtodg;
  q_b = q_b_ic_dg/rdtodg;
}

void kinematics() {
  // This subroutine determined the kinematic state derivatives.
  //
  // Inputs:
  //     AaeroX_bi_b - Accelerations due to aero in BCS [mXsec**2)
  //     AaeroZ_hi_b
  //     Gravity - Acceleration due to gravity [m/sec**2]
  //     Theta_b - Attitude of missile [rad]
  //
  // Outputs:
  //     Xdd_bi_i - X acceleration of BCS w.r.t. ICS, expressed in the ICS [m/sec]
  //     Zdd_bi_i - Z acceleration of BCS w.r.t. ICS expressed in the ICS [m/sec]
  // Internal variables and constants:
  //
  // Begin math model:
  // Calculate velocities (attitude rate comes directly from airframe response)
  xdd_bi_i =  aaerox_bi_b*cos(theta_b) + aaeroz_bi_b*sin(theta_b);
  zdd_bi_i = -aaerox_bi_b*sin(theta_b) + aaeroz_bi_b*cos(theta_b)+ gravity;
}
